### PR TITLE
add mobile-client-config component

### PR DESCRIPTION
**Pull request by philbrookes**
Add config-editor component with a copy to clipboard function:
screen shot 2018-03-23 at 09 31 01

screen shot 2018-03-23 at 09 31 08

### CODE REVIEW

File: app/scripts/directives/configEditor.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+angular.module('openshiftConsole')
+  .directive('configEditor', function(APIService, DataService) {
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: If you can use the `copyToClipBoard` directive I would move this to a component.

File: app/views/config-editor.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<div class="component-label section-label">Config Info
+  <button class="clipboard-button pull-right btn btn-default" data-clipboard-text="{{prettyConfig}}" ng-click="copy()">Copy Config</button>
+</div>
+<div ui-ace="{
```
Comment (sedroche) [inline app/views/config-editor.html]: Can these be set on the controller?

File: app/views/config-editor.html [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<div class="component-label section-label">Config Info
+  <button class="clipboard-button pull-right btn btn-default" data-clipboard-text="{{prettyConfig}}" ng-click="copy()">Copy Config</button>
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply app/views/config-editor.html]: no, that doesn't allow you to edit the contents of the textarea unfortunately.

File: app/scripts/directives/configEditor.js (hunk 1 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: I cannot use the copyToClipBoard directive.

File: app/views/config-editor.html (hunk 2 above)
Comment (philbrookes) [reply app/views/config-editor.html]: I did it this way, as this is how it is done elsewhere in the code-base. e.g.
https://github.com/aerogear/origin-web-console/blob/master/app/views/browse/build-config.html#L339

**Comment by sedroche** [comment]
@philbrookes I would separate this directive into 2 parts.

1. A directive that is just a button that copies the supplied text to clipboard. The directive should only manange html state.
2. A parent component that gets the config and manages the text and passes it to the directive.

That way we have a more reusable copy to clipboard directive.

File: app/scripts/directives/configEditor.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: `mobileConfigEditor` as the name?

I would change to `.component` now as we don't use `link`

File: app/scripts/directives/configEditor.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    return {
+      restrict: 'E',
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: I think this can be removed in components too.

File: app/scripts/directives/configEditor.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      scope: {
+        mobileClient: '<',
+      },
+      templateUrl: 'views/config-editor.html',
+      controller: function($scope) {
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: Unless you specifically need `$scope` I would use `$ctrl`

File: app/scripts/directives/configEditor.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        $scope.context = {namespace: _.get($scope, "mobileClient.metadata.namespace")};
+        //keep list of active $scope.services upto date
+        DataService.watch(
+          APIService.getPreferredVersion('serviceinstances'), 
+          $scope.context, 
+          function (serviceinstances){
+            $scope.services = [];
+            _.each(serviceinstances.by("metadata.name"), function(service, serviceName){
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: Does `_.filter` suit better here?

File: app/scripts/directives/configEditor.js (hunk 1 above)
Comment (maleck13) [reply app/scripts/directives/configEditor.js]: Why can't it be used?

File: app/scripts/directives/configEditor.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+              if(_.get(service, "metadata.labels.mobile") !== "enabled") {
```
Comment (maleck13) [inline app/scripts/directives/configEditor.js]: As we soon to remove labeling the service instance, does this make sense? I think we should be just looking for the configmaps labelled mobile enabled?

File: app/scripts/directives/configEditor.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+          { errorNotification: false }
+        );
+        DataService.watch(APIService.getPreferredVersion('configmaps'), $scope.context, updateConfig, {errorNotification: false});
+
+        // update the config string by pulling out configmaps that match $scope.services
+        function updateConfig(configmaps){
+          var configs = [];
+          _.each(configmaps.by("metadata.name"), function(configmap, configmapName){
+            if(_.indexOf($scope.services, configmapName) !== -1){
+              console.log(configmap);
```
Comment (maleck13) [inline app/scripts/directives/configEditor.js]: remove console log

File: app/scripts/directives/configEditor.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+              }
+              if(_.indexOf($scope.mobileClient.spec.excludedServices, serviceName) === -1){
+                $scope.services.push(_.get(service, "metadata.labels.serviceName"));
+              }
+            });
+            DataService.list(APIService.getPreferredVersion('configmaps'), $scope.context, updateConfig, {errorNotification: false});
+          }, 
+          { errorNotification: false }
+        );
+        DataService.watch(APIService.getPreferredVersion('configmaps'), $scope.context, updateConfig, {errorNotification: false});
```
Comment (maleck13) [inline app/scripts/directives/configEditor.js]: Is it possible to add a label selector here? All the configs we are interested in will be labelled mobile:enabled

File: app/scripts/directives/configEditor.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+        // update the config string by pulling out configmaps that match $scope.services
+        function updateConfig(configmaps){
+          var configs = [];
+          _.each(configmaps.by("metadata.name"), function(configmap, configmapName){
+            if(_.indexOf($scope.services, configmapName) !== -1){
```
Comment (maleck13) [inline app/scripts/directives/configEditor.js]: Again I don't think we need this if we filter on mobile:enabled

**Comment by maleck13** [comment]
Also could you add a set of verification steps to the PR. Ping @AdamSaleh

File: app/scripts/directives/configEditor.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                name: configmapName,
+                type: configmap.data.type,
+                url: configmap.data.uri,
+                config: configmap.data
+              });
+            }
+          });
+          var ret = {
+            "version": 1,
+            "clusterName": window.OPENSHIFT_CONFIG.apis.hostPort,
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: Is this a requirable constant?

File: app/scripts/directives/util.js [hunk 13]
```diff
@@ -103,6 +103,36 @@ angular.module('openshiftConsole')
...
   })
+  .directive('copyToClipboardButton', function(NotificationsService){
+    return {
+      restrict: 'E',
+      replace: true,
+      scope: {
+        clipboardText: "@",
+        buttonText: "@"
+      },
+      template: '<button class="clipboard-button pull-right btn btn-default" data-clipboard-text="{{clipboardText}}">{{buttonText}}</button>',
```
Comment (sedroche) [inline app/scripts/directives/util.js]: `pull-right` should be up to the user of the directive

File: app/scripts/directives/util.js [hunk 14]
```diff
@@ -103,6 +103,36 @@ angular.module('openshiftConsole')
...
+          $(e.trigger)
+            .attr('title', 'Copied!')
+            .tooltip('fixTitle')
+            .tooltip('show')
+            .attr('title', 'Copy to Clipboard')
+            .tooltip('fixTitle');
+          e.clearSelection();
+        });
+        clipboard.on('error', function () {
+          NotificationsService.addNotification({
```
Comment (sedroche) [inline app/scripts/directives/util.js]: `onError` binding that the user of the directive decides how to respond might be better here.

**Comment by AdamSaleh** [comment]
@philbrookes w.r.t. the verification steps, would something like this suffice?

Steps:
1. Create a new project
2. Create an Android mobile client
3. View there is a Config Info section
  - in the mobile-clients representation screen
  - in the mobile client section of the overview screen 
  - the copy-button on all the screen works
4. Create a mobile service (i.e. fh-sync) in a separate screen/through cli
5. Verify that the Config Info in the screen has changed accordingly
  - in the mobile-clients representation screen
  - in the mobile client section of the overview screen 
6. Add another configmap with the "with the mobile:enabled label"
7. Verify as described in step 5
8. Edit the configmap
7. Verify as described in step 5
9. Delete the configmap
7. Verify as described in step 5

I am basing this on the jira: https://issues.jboss.org/browse/AEROGEAR-2270

If you think this would be sufficient can you add it as the verification steps?

**Comment by philbrookes** [comment]
@AdamSaleh 

Verification steps:
1. create mobile client, see empty configuration
2. provision fh-sync-server - configuration editor should live-update without a refresh required
3. confirm copy button copies values to paste buffer
5. deprovision fh-sync-server - config-editor should live-update

As far as I know, there is no mobile client representation screen yet, to test on, I tested this using the overview page

File: app/scripts/directives/configEditor.js (hunk 1 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: It cannot be used without displaying a textarea, and the textarea cannot be editable.
If this field does not need to be editable, then I could re-use it, but I recall a conversation where it was stressed that this should be editable?

File: app/scripts/directives/configEditor.js (hunk 8 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: I have to scan through services first, to ignore any excluded ones, so I have moved the check for this label to the configmap watch function, and removed it from here.

File: app/scripts/directives/configEditor.js (hunk 6 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: when I change this to `$ctrl` the prettyConfig is no longer available on the template.

File: app/scripts/directives/configEditor.js (hunk 7 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: yes, updated this now.

File: app/scripts/directives/configEditor.js (hunk 10 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: I don't see a way of filtering this, other than in the callback

File: app/scripts/directives/configEditor.js (hunk 11 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: the point of this is to check that the configmap is not an excluded service.

File: app/scripts/directives/configEditor.js (hunk 4 above)
Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: if I change this to `.component` it stops working

**Comment by philbrookes** [comment]
@AdamSaleh @JameelB this is ready for verification now.

**Comment by philbrookes** [comment]
@AdamSaleh I discussed this PR with Craig over bluejeans, and for now the requirement for the serviceInstance to be running will have to remain.

File: app/scripts/directives/configEditor.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+
+angular.module('openshiftConsole').component('configEditor', {
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: I would still make this a `mobileConfigEditor` as it is specific to mobile

File: app/scripts/directives/configEditor.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        DataService.list(APIService.getPreferredVersion('configmaps'), context, updateConfig, {errorNotification: false});
+      }, 
+      { errorNotification: false }
+    );
+    watches.push(DataService.watch(APIService.getPreferredVersion('configmaps'), context, updateConfig, {errorNotification: false}));
+
+    // update the config string by pulling out configmaps that match $scope.services
+    function updateConfig(configmaps){
+      var configs = [];
+      _.each(configmaps.by("metadata.name"), function(configmap, configmapName){
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: I was looking at Lodash and I believe you can do this in a bit more functional style:

```.js
var configs =_(configmaps.by("metadata.name"))
  .filter(function(configmap) {
   return  _.findIndex(ctrl.services, {metadata: {labels: {serviceName: configmapName}}}) !== -1;
  })
  .map(function(configmap) {
   return {
       id: configmapName,
       name: configmapName,
       type: configmap.data.type,
       url: configmap.data.uri,
       config: configmap.data
   };
  })
  .value()
```

File: app/scripts/directives/configEditor.js [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+      'API_CFG', 
+      ConfigEditorCtrl]
+  });
+
+function ConfigEditorCtrl(APIService, DataService, API_CFG) {
+  var ctrl = this;
+  var watches = [];
+  ctrl.$onInit = function(){
+    var context = {namespace: _.get(ctrl, "mobileClient.metadata.namespace")};
+    //keep list of active $scope.services upto date
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: Remove the `$scope` reference

File: app/views/overview/_mobile-client-row.html [hunk 18]
```diff
@@ -37,18 +37,11 @@ <h3>
   </div>
   <div class="list-pf-expansion collapse" ng-if="row.expanded" ng-class="{ in: row.expanded }">
     <div class="list-pf-container">
-      <div class="expanded-section">
-        <div class="empty-state-message text-center">
-          <p>Add a mobile service to your project. Or connect to external service.</p>
-          <div class="empty-state-message-main-action">
-            <!--TODO link through to the mobile category -->
-            <button class="btn btn-primary btn-lg" ng-click="row.browseCatalog()">
```
Comment (sedroche) [inline app/views/overview/_mobile-client-row.html]: I think we want to keep this as part of the component

File: app/scripts/directives/configEditor.js [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+              name: configmapName,
+              type: configmap.data.type,
+              url: configmap.data.uri,
+              config: configmap.data
+            });
+          }
+        }
+      });
+      var ret = {
+        "version": 1,
```
Comment (sedroche) [inline app/scripts/directives/configEditor.js]: Is this always `1`?

Comment (philbrookes) [reply app/scripts/directives/configEditor.js]: it is hardcoded as 1 in the mobile-cli, where I copied it from

**Comment by philbrookes** [comment]
opening upstream

File: app/scripts/directives/mobileClientConfig.js [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+          config: configmap.data
+        };
+      }).value();
+      ctrl.prettyConfig = JSON.stringify({
+        version: 1,
+        clusterName: API_CFG.openshift.hostPort,
+        namespace: _.get(ctrl, "mobileClient.metadata.namespace"),
+        clientId: _.get(ctrl, "mobileClient.metadata.name"),
+        services: configs
+      }, null, "  ");
```
Comment (vchepeli) [inline app/scripts/directives/mobileClientConfig.js]: should not it be number `JSON.stringify(text, null, 2)`? see you have two spaces here

File: app/views/overview/_mobile-client-row.html [hunk 21]
```diff
@@ -37,18 +37,21 @@ <h3>
...
+        </div>
+        <div class="col-md-7">
+        </div>
+        <div class="col-md-12">
           <p>Add a mobile service to your project. Or connect to external service.</p>
           <div class="empty-state-message-main-action">
-            <!--TODO link through to the mobile category -->
-            <button class="btn btn-primary btn-lg" ng-click="row.browseCatalog()">
-              Browse Mobile Services
+            <button class="btn btn-primary btn-lg" ng-click="row.browseCatalog()">Browse Mobile Services
```
Comment (vchepeli) [inline app/views/overview/_mobile-client-row.html]: Better to move `Browse Mobile Services` to next line or move `</button>` to the end of this

**Comment by vchepeli** [comment]
@AdamSaleh, @maleck13 you verified these changes guys locally?

**Comment by vchepeli** [comment]
ANDROID
```js
{
  "version": 1,
  "clusterName": "192.168.64.47:8443",
  "namespace": "test1",
  "clientId": "myapp-android",
  "services": [
    {
      "id": "fh-sync-server",
      "name": "fh-sync-server",
      "type": "fh-sync-server",
      "url": "https://fh-sync-server-test1.192.168.64.47.nip.io",
      "config": {
        "name": "fh-sync-server",
        "type": "fh-sync-server",
        "uri": "https://fh-sync-server-test1.192.168.64.47.nip.io"
      }
    },
    {
      "id": "keycloak",
      "name": "keycloak",
      "type": "keycloak",
      "url": "https://keycloak-test1.192.168.64.47.nip.io",
      "config": {
        "bearer_client_id": "yupgAOMnlfMhuHfAauWb-bearer",
        "bearer_client_secret": "6PzwNcQDT60ExW6Fuqgy",
        "bearer_installation": "{\"realm\": \"test1\", \"ssl-required\": \"external\", \"confidential-port\": 0, \"auth-server-url\": \"https://keycloak-test1.192.168.64.47.nip.io/auth\", \"bearer-only\": true, \"resource\": \"yupgAOMnlfMhuHfAauWb-bearer\"}",
        "name": "keycloak",
        "public_client_id": "yupgAOMnlfMhuHfAauWb",
        "public_client_secret": "6PzwNcQDT60ExW6Fuqgy",
        "public_installation": "{\"realm\": \"test1\", \"ssl-required\": \"external\", \"clientId\": \"yupgAOMnlfMhuHfAauWb\", \"auth-server-url\": \"https://keycloak-test1.192.168.64.47.nip.io/auth\", \"url\": \"https://keycloak-test1.192.168.64.47.nip.io/auth\", \"resource\": \"yupgAOMnlfMhuHfAauWb\"}",
        "realm": "test1",
        "type": "keycloak",
        "uri": "https://keycloak-test1.192.168.64.47.nip.io"
      }
    }
  ]
}
```

IOS
```js
{
  "version": 1,
  "clusterName": "192.168.64.47:8443",
  "namespace": "test1",
  "clientId": "myapp-ios",
  "services": [
    {
      "id": "fh-sync-server",
      "name": "fh-sync-server",
      "type": "fh-sync-server",
      "url": "https://fh-sync-server-test1.192.168.64.47.nip.io",
      "config": {
        "name": "fh-sync-server",
        "type": "fh-sync-server",
        "uri": "https://fh-sync-server-test1.192.168.64.47.nip.io"
      }
    },
    {
      "id": "keycloak",
      "name": "keycloak",
      "type": "keycloak",
      "url": "https://keycloak-test1.192.168.64.47.nip.io",
      "config": {
        "bearer_client_id": "yupgAOMnlfMhuHfAauWb-bearer",
        "bearer_client_secret": "6PzwNcQDT60ExW6Fuqgy",
        "bearer_installation": "{\"realm\": \"test1\", \"ssl-required\": \"external\", \"confidential-port\": 0, \"auth-server-url\": \"https://keycloak-test1.192.168.64.47.nip.io/auth\", \"bearer-only\": true, \"resource\": \"yupgAOMnlfMhuHfAauWb-bearer\"}",
        "name": "keycloak",
        "public_client_id": "yupgAOMnlfMhuHfAauWb",
        "public_client_secret": "6PzwNcQDT60ExW6Fuqgy",
        "public_installation": "{\"realm\": \"test1\", \"ssl-required\": \"external\", \"clientId\": \"yupgAOMnlfMhuHfAauWb\", \"auth-server-url\": \"https://keycloak-test1.192.168.64.47.nip.io/auth\", \"url\": \"https://keycloak-test1.192.168.64.47.nip.io/auth\", \"resource\": \"yupgAOMnlfMhuHfAauWb\"}",
        "realm": "test1",
        "type": "keycloak",
        "uri": "https://keycloak-test1.192.168.64.47.nip.io"
      }
    }
  ]
}
```

**Comment by vchepeli** [comment]
@maleck13, @AdamSaleh verified @philbrookes changes locally

**Pull request closed without merging**